FastText/src/models/custom_layers.jl

```diff
@@ -0,0 +1,410 @@
+  (m.active = (isnothing(mode) || mode == :auto) ? nothing : !mode; m)
+
+function reset_masks!(de::DroppedEmbeddings)
+    de.mask = drop_mask(de.mask, de.p)
+    return
+end
+####################################################################
+
+################# Concat Pooling Dense layer #######################
+"""
+    PooledDense(hidden_sz::Integer, out::Integer, σ = identity)
+
+Concat-Pooled Dense layer
+
+This is basically a modified version of the `Dense` layer.
+It takes the `Vector` of outputs of RNN at all time-steps,
+then it calculates the mean and max pools for those outputs and
+concatenates output RNN at the last time-step with these max and mean pools.
+Then this conatenated `Vector` is multiplied with weights and added with bias
+and passes through specified activation function.
+
+Usage:
+The first argument `hidden_sz` takes length of the ouput of the preceding RNN layer.
+Other two arguments are output size and activation function
+
+# Example
+
+julia> pd = PooledDense(40, 20)    # if the output size of the RNN layer is 40 in this case
+"""
+mutable struct PooledDense{F, S, T}
+    W::S
+    b::T
+    σ::F
+end
+
+PooledDense(W, b) = PooledDense(W, b, identity)
+
+function PooledDense(hidden_sz::Integer, out::Integer, σ = identity;
+             initW = Flux.glorot_uniform, initb = (dims...) -> zeros(Float32, dims...))
+return PooledDense(initW(out, hidden_sz*3), initb(out), σ)
+end
+
+Flux.@functor PooledDense
+
+function (a::PooledDense)(x)
+    W, b, σ = a.W, a.b, a.σ
+    x = cat(x..., dims=3)
+    maxpool = maximum(x, dims=3)[:, :, 1]
+    meanpool = (sum(x, dims=3)/size(x, 3))[:, :, 1]
+    hc = cat(x[:, :, 1], maxpool, meanpool, dims=1)
+    σ.(W*hc .+ b)
+end
+
+####################################################################
+
+"""
+get_trainable_params(layers)
+
+This funciton works same as `params` function except for `AWD_LSTM` layer.
+While getting `Params` of the `AWD_LSTM` it does not include the `h` and `c` `params` of `AWD_LSTM`.
+This is useful while calculating gradients because calculating gradients for `h` and `c` fields
+in `AWD_LSTM` is unnecessary here.
+
+# Example:
+
+julia> layers = Chain(DroppedEmbeddings(4,5,0.2),
+                    AWD_LSTM(5, 3),
+                    Dense(3, 2),
+                    softmax
+                );
+julia> p1 = params(layers);
+julia> p2 = get_trainable_params(layers);
+
+julia> length(p1)
+8
+
+julia> length(p2)
+6
+
+`Params` from all the other layers are included in p2 except for `h` and `c`
+"""
+function get_trainable_params(layers)
+    p = []
+    function get_awd_params(awd::AWD_LSTM)
+        return [awd.layer.cell.Wi,
+        awd.layer.cell.Wh,
+        awd.layer.cell.b]
+    end
+    for layer in layers
+        layer isa Array || (layer = [layer])
+        for l in layer
+            l isa AWD_LSTM && (append!(p, get_awd_params(l)); continue)
+            push!(p, l)
+        end
+    end
+    return Flux.params(p...)
+end
```

Comment:
I would get rid of this. If it's needed for the training code, we can discuss how to replace it.

Reply:
That won't be necessary right now, I'll remove this.